Add the delete action

ekinanp opened the pull request:
Subsequent PRs will implement Delete for the core plugins.

ekinanp committed:
Escape the path in plugin.opKeysRegex

This commit also updates plugin.ClearCacheFor to no longer return an
error now that opKeysRegex is guaranteed to always produce a valid
regex for any given path.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add plugin.EntryMap

We'll need this type to properly implement Delete.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the Delete action

This commit also updates plugin.List's return type to plugin.EntryMap
and makes the appropriate modifications to reflect that change.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the /fs/delete endpoint

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the delete command

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement external plugin delete

This makes it easy to test the new Delete action.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Panic if plugin root implements delete

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
I tested this with the `local_fs` plugin; its delete implementation delegated to `rm -rf`. Things seem to be working OK.

ekinanp commented on cmd/rm.go:
```diff
@@ -0,0 +1,69 @@
+		Args: cobra.MinimumNArgs(1),
+		RunE: toRunE(rmMain),
+	}
+	rmCmd.Flags().BoolP("recurse", "r", false, "Delete directories (parent entries)")
```
Hmm, given the asymmetry between the `fs/delete` endpoint and the `Delete` method's signature (something that isn't an issue in our other actions), I'm starting to think that maybe the `recurse` option isn't such a good idea. Is there a compelling reason for us to enforce it? It seems simple enough to just say "`wrm` removes the given entries" (or, equally, that "rm"'s semantics don't cleanly apply to Wash, which is a good reason to change `wrm` => `delete` so that people don't get confused).

(If we talked about not having `-r` on Wednesday's meeting and I spoke against it, then apologies for that).

MikaelSmith replied:
Updated the issue description, I think it's clear to remove `-r` and rename the command.

ekinanp commented on plugin/types.go:
```diff
@@ -193,3 +193,12 @@ type Readable interface {
+// error.
+type Deletable interface {
+	Entry
+	Delete(context.Context) error
```
Deleting an EC2 instance is equivalent to terminating it. According to https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/terminating-instances.html, termination is irreversible. However, terminated EC2 instances do show up in the AWS console for an hour.

Reason I bring this up is b/c if we delete an EC2 instance in Wash, that EC2 instance will disappear from the `instances` directory (breaking consistency with the AWS console). Is that OK? If so, we should probably filter out terminated EC2 instances when listing the `instances` directory.

ekinanp replied:
I lean towards not implementing `Delete` for EC2 instances and maybe outlining the reason in `docs <ec2_instance>`.

MikaelSmith replied:
I feel like this might be a common behavior. Deleting things may not be instantaneous, and we need to come up with some way to reflect that they're being removed.

MikaelSmith replied:
I think EC2 should respond to Delete, same as GCP. In each case, it looks like they may go through a TERMINATED state before being removed. I think for now, we should consider TERMINATED as absent and not list those instances. The Delete implementation for them should only return once they've arrived at the TERMINATED state.

The API should trigger Delete on batched instances in parallel so that waiting on multiple to go through that lifecycle doesn't take too long.

MikaelSmith commented on plugin/methodWrappers_test.go:
```diff
@@ -67,19 +86,80 @@ func newmethodWrappersTestsMockEntry(name string) *methodWrappersTestsMockEntry
+	expectedErr := fmt.Errorf("an error")
+	e.On("Delete", ctx).Return(expectedErr)
+
+	var entry Entry = e
```
Why the cast to `Entry`?

ekinanp committed:
Use e.id() instead of ID(e) in plugin.Delete

This is consistent with what we do in the other code inside the plugin
package.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Remove unnecessary entry variable in plugin.Delete tests

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add a concurrent read/write test to plugin.EntryMap

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on plugin/entryMap_test.go:
```diff
@@ -0,0 +1,140 @@
+	suite.True(count <= 1)
+}
+
+func (suite *EntryMapTestSuite) TestConcurrentReadWrite() {
```
This test does panic if any of Load/Delete/Range fail to acquire a lock or release an acquired lock. It also panics if Delete acquires a read lock instead of a write lock.

ekinanp commented:
Updated to address the PR feedback.

MikaelSmith commented:
Initial work for https://github.com/puppetlabs/wash/issues/479.
